### ISSUE TITLE
Scrolls are not sufficiently cleared on AWS Opensearch: rework fix for #19869

### ISSUE THREAD

**Issue by ThorbenLindhauer** [camunda/camunda#24321]
### Describe the bug

* With https://github.com/camunda/camunda/issues/24262 we plan to revert the change of https://github.com/camunda/camunda/issues/19869, namely we want to move back to an Apache-Http-Client-based AWS transport.
* This introduces the problem from https://github.com/camunda/camunda/issues/24262 again as a regression that needs to be solved

### To Reproduce

1. Make an Opensearch request through Operate that uses the Scroll API
1. Observe a log warning similar to below

```
2024-11-04 13:45:23 2024-11-04 12:45:23.867 [] [import_1] [] WARN 
2024-11-04 13:45:23       io.camunda.operate.store.opensearch.client.sync.RichOpenSearchClient - Error occurred when clearing the scroll with id [FGluY2x1ZGVfY29udGV4dF91dWlkDnF1ZXJ5VGhlbkZldGNoAhZzN3ZHc3lSMlFrYTRJWmFiMHdNYVFBAAAAAAAAOEIWcE92d3lwZ0JSbk9CXzJtaVdWR2Y1ZxZzN3ZHc3lSMlFrYTRJWmFiMHdNYVFBAAAAAAAAOEMWcE92d3lwZ0JSbk9CXzJtaVdWR2Y1Zw==]
```

### Current behavior

* The scroll object cannot be successfully cleared and remains open
* We make scroll requests with a timeout of 1 minute: https://github.com/camunda/camunda/blob/8.6.4/operate/schema/src/main/java/io/camunda/operate/store/opensearch/client/sync/OpenSearchDocumentOperations.java#L52
  * Opensearch cleans up any open scrolls after this time
  * Should this not be specified, the default value is 5 minutes; see config `search.default_keep_alive` in [Opensearch reference](https://opensearch.org/docs/latest/install-and-configure/configuring-opensearch/search-settings/)
* While the scroll is not yet removed, it consumes memory in Opensearch and counts towards the max scroll limit (default: 500); see config `search.max_open_scroll_contex` in [Opensearch reference](https://opensearch.org/docs/latest/install-and-configure/configuring-opensearch/search-settings/)
* Once the max scroll limit is reached, new scroll requests are rejected (see exception in https://github.com/camunda/camunda/issues/19869)

### Expected behavior

* Scroll objects are successfully cleared when using AWS Opensearch

### Environment

- Operate 8.6/85./8.4/8.3 with latest patches (as of writing) and AWS Opensearch

### Additional context

* https://jira.camunda.com/browse/SUPPORT-23870
* https://jira.camunda.com/browse/SUPPORT-24097
* https://jira.camunda.com/browse/SUPPORT-24772
* https://jira.camunda.com/browse/SUPPORT-24625
* [SUPPORT-24985](https://jira.camunda.com/browse/SUPPORT-24985)
* [SUPPORT-25126](https://jira.camunda.com/browse/SUPPORT-25126)

---------------------------------------------------------------------------------------------

### Acceptance Criteria

* Operate does not fail to clear scroll objects when used with AWS Opensearch (e.g. the log message above is not logged anymore)
* The Operate QA IT test suite succeeds against Opensearch

### Definition of Ready - Checklist

- [ ] The bug has been reproduced by the assignee in the environment compatible with the provided one; otherwise, the issue is closed with a comment
- [ ] The issue has a meaningful title, description, and testable acceptance criteria
- [ ] The issue has been labeled with an appropriate `Bug-area` label
- [ ] Necessary screenshots, screen recordings, or files are attached to the bug report

For UI changes required to solve the bug:

- [ ] Design input has been collected by the assignee

#### PRs

* main: https://github.com/camunda/camunda/pull/26273
* 8.6: https://github.com/camunda/camunda/pull/26268
* 8.5:
  * Operate: https://github.com/camunda/camunda/pull/26272
  * Tasklist: https://github.com/camunda/tasklist/pull/5097
* 8.4:
  * Operate: https://github.com/camunda/operate/pull/6833
  * Tasklist: https://github.com/camunda/tasklist/pull/5098
* 8.3:
  * Operate: https://github.com/camunda/operate/pull/6834

#### :mag: Root Cause Analysis

* When [we make a clear scroll request](https://github.com/camunda/camunda/blob/8.6.4/operate/schema/src/main/java/io/camunda/operate/store/opensearch/client/sync/OpenSearchDocumentOperations.java#L75-L83) in this setup, it fails with the following exception message:
```
org.opensearch.client.opensearch._types.OpenSearchException: Request failed: [security_exception] The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details.

The Canonical String for this request should have been
'DELETE
/_search/scroll

accept-encoding:gzip
content-length:
content-type:application/json
host:<redacted Opensearch URL>
x-amz-content-sha256:<redacted signature of the request>
x-amz-date:20241104T104947Z

accept-encoding;content-length;content-type;host;x-amz-content-sha256;x-amz-date
<redacted>'

The String-to-Sign should have been
<redacted>
```
* The reason is that the Opensearch Java client includes the Content-Length header in the signature calculation, but the underlying Apache HttpClient doesn't send the header
  * The Opensearch AWS Transport implementation calculates a Content-Length header for any request: https://github.com/opensearch-project/opensearch-java/blob/v2.9.0/java-client/src/main/java/org/opensearch/client/transport/aws/AwsSdk2Transport.java#L319
  * Only after the previous step, it signs the request including all request headers: https://github.com/opensearch-project/opensearch-java/blob/v2.9.0/java-client/src/main/java/org/opensearch/client/transport/aws/AwsSdk2Transport.java#L350
  * The AWS SDK code that transforms an AWS request into an Apache request omits the Content-Length header: https://github.com/aws/aws-sdk-java-v2/blob/2.29.6/http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/impl/ApacheHttpRequestFactory.java#L175-L179
  * The Apache Http client calculates the content length header generally on its own: https://github.com/apache/httpcomponents-core/blob/rel/v4.4.16/httpcore/src/main/java/org/apache/http/protocol/RequestContent.java#L114
  * But it only does that for requests that are supposed to have a request body: https://github.com/apache/httpcomponents-core/blob/rel/v4.4.16/httpcore/src/main/java/org/apache/http/protocol/RequestContent.java#L88
  * `HttpDelete` (which is used for this request) is not such a request => This explains why other HTTP methods like POST are not affected by the problem
    * Side note: The Opensearch core rest client has a custom implementation of this Apache class that allows for this case: https://github.com/opensearch-project/OpenSearch/blob/2.9.0/client/rest/src/main/java/org/opensearch/client/HttpDeleteWithEntity.java. It's not however not easily possible to plug this in where we'd need it. It's also removed in newer versions where the rest client was migrated from Apache Http client 4 to 5
    * Apache Http client 5 doesn't have this problem, because it allows for deletes with request bodies: https://issues.apache.org/jira/browse/HTTPCORE-380
  * In effect, no content length header is sent to the server, but it includes this empty header value in the process to validate the request signature; this leads to mismatching signatures and a rejected request

#### :thought_balloon: Proposed Solution

* Wait to get https://github.com/opensearch-project/opensearch-java/pull/1256 and the follow-up https://github.com/aws/aws-sdk-java-v2/pull/5704 merged
* Update to an Opensearch Java client version that contains the fix
* Alternative: Switch to yet another http client implementation, e.g. `AwsCrtHttpClient`: https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/http-configuration.html
  * However, this may come with its own set of challenges that we are not anticipating and may introduce new regressions

### :point_right: Handover Dev to QA

- Changed components:
- Side effects on other components:
- Handy resources:
  BPMN/DMN models, plugins, scripts, REST API endpoints + example payload, etc :
  
- Example projects:

- Commands/Steps needed to test; Versions to validate:

- Docker file / HELM chart : in case that it needed to be tested via docker share the version contain the fixed along with version of other services .

- Release version ( in which version this fixed/feature will be released):

**Comment by victorekpo** [comment]
Thank you for breaking down this issue! I couldn't figure out why the requests were failing with the Sdk2Transport

**Comment by ThorbenLindhauer** [comment]
@houssain-barouni note that there is a follow up PR for the AWS Java SDK to support DELETEs with request bodies: https://github.com/aws/aws-sdk-java-v2/pull/5704

Please check if we need this fix so that this works correctly. In my high level-understanding we do. See also the contributor's comment here: https://github.com/opensearch-project/opensearch-java/pull/1256#issuecomment-2464320396

**Comment by houssain-barouni** [comment]
@ThorbenLindhauer [aws-java-sdk 2.29.22](https://github.com/aws/aws-sdk-java-v2/releases/tag/2.29.22). The dependency was upgraded by Renovate on main branch https://github.com/camunda/camunda/pull/25326. I will test it when I will have some bandwidth this week.

**Comment by houssain-barouni** [comment]
we also need this https://github.com/opensearch-project/opensearch-java/pull/1333 to be merged and released

**Comment by ThorbenLindhauer** [comment]
Thanks for staying on top of this, Houssain.

**Comment by alexcorremans** [comment]
Support tickets:
https://jira.camunda.com/browse/SUPPORT-23870
https://jira.camunda.com/browse/SUPPORT-24097
https://jira.camunda.com/browse/SUPPORT-24772
https://jira.camunda.com/browse/SUPPORT-24773

Should the `support` label be added to this github issue?

**Comment by ThorbenLindhauer** [comment]
Thanks, @alexcorremans. I have added the links to the ticket description so that the release notification automtaion should pick it up.

**Comment by houssain-barouni** [comment]
@ThorbenLindhauer some bad news, the fix in aws-sdk-java got reverted because of some side effects (cf https://github.com/aws/aws-sdk-java-v2/releases/tag/2.29.26). This may take a while to be fixed again.

**Comment by alexcorremans** [comment]
https://jira.camunda.com/browse/SUPPORT-24773 should probably be removed from the description, they don't seem to be experiencing this bug after all.

**Comment by ThorbenLindhauer** [comment]
@houssain-barouni here is the new PR to follow: https://github.com/aws/aws-sdk-java-v2/pull/5743

**Comment by ThorbenLindhauer** [comment]
Over the weekend, I ran the Operate ITs against AWS Opensearch, using the Amazon CRT client instead of Apache. Results attached. Will next check individual test failures if they persist, similar to what I did for https://github.com/camunda/camunda/issues/24262#issuecomment-2457929019. This would potentially allows us to make progress with this independent of an external fix.

[24321-build-log.txt](https://github.com/user-attachments/files/18187969/24321-build-log.txt)

**Comment by ThorbenLindhauer** [comment]
I think we can move ahead with the CRT client based on the Operate integration tests.

Verified that the the scroll bug is fixed by running `BasicZeebeImportIT#testIncidentCreatesProcessInstance`: 

* produces the logs `Error occurred when clearing the scroll with id [..]` with Apache cleint
* doesn't show them with the CRT client

Analysis of test failures with CRT client (based on log above)

* DecisionIT
  * succeeds when run in isolation (like [before](https://github.com/camunda/camunda/issues/24262#issuecomment-2457929019))
* SchemaCreationIT
  * failed because I had set an environment variable for the default number of replica shards in my test setup, which the test didn't expect
  * variables were set to 1 primary shard, 2 replicas, i.e. one shard per OS node in my setup
  * When removing them, I received an exception as follows `org.opensearch.client.opensearch._types.OpenSearchException: Request failed: [invalid_index_template_exception] index_template [fb81335d-b-operate_template] invalid, cause [Validation Failed: 1: expected total copies needs to be a multiple of total awareness attributes [3]`
  * I believe this is expected behavior with my AWS Opensearch setup; the application is not broken here, just the test not adapted to my setup
* SchemaManagerIT
  * The test doesn't work with Opensearch in general; the test asserts that an index mapping's meta properties are null; but the Opensearch implementation initializes it as an empty hash map (io.camunda.operate.store.opensearch.client.sync.OpenSearchIndexOperations.getIndexMappings(String)); I reviewed the code and don't see a concrete problem here; note that this affects both AWS and regular Opensearch
* BackupControllerIT
  * Test is specific to Elasticsearch (mocks responses of the ES client and verifies ES client interactions)
* ElasticsearchConnectorIT
  * Specific to Elasticsearch
* ProcessStoreIT
  * looks like a general Opensearch regression; not specific to the client change; unlike the Elasticsearch implementation, the Opensearch code concatenates tenant id and bpmn process id as the key in the result map (see io.camunda.operate.store.opensearch.OpensearchProcessStore.getProcessesGrouped(String, Set<String>))
* AuthenticationWithPersistentSessionsIT
  * Tests are specific to Elasticsearch (loads only the ES-specific application context classes)
* FlowNodeMetadataZeebeImportIT
  * https://github.com/camunda/camunda/issues/24397
* ImportMidnightZeebeImportIT
  * succeeds when run in isolation
* ProcessInstanceZeebeImportIT
  * the failing test succeeds in isolation

**Comment by ThorbenLindhauer** [comment]
How to run the Operate integration tests agains AWS Opensearch:

I pretty much hacked this in as the current Operate IT code is not extensible for this.

Here is what I did:

In the integration tests pom, I set the following [env vars for the failsafe plugin](https://github.com/camunda/camunda/blob/8.6.4/operate/qa/integration-tests/pom.xml#L479-L481):

```
<CAMUNDA_OPERATE_DATABASE>opensearch</CAMUNDA_OPERATE_DATABASE>
<CAMUNDA_OPERATE_OPENSEARCH_URL>${opensearch-url}</CAMUNDA_OPERATE_OPENSEARCH_URL>
<CAMUNDA_OPERATE_ZEEBEOPENSEARCH_URL>${opensearch-url}</CAMUNDA_OPERATE_ZEEBEOPENSEARCH_URL>
<AWS_REGION>${region}</AWS_REGION>
<AWS_ACCESS_KEY_ID>${accessKey}</AWS_ACCESS_KEY_ID>
<AWS_SECRET_ACCESS_KEY>${secretKey}</AWS_SECRET_ACCESS_KEY>
<CAMUNDA_OPERATE_OPENSEARCH_NUMBEROFSHARDS>1</CAMUNDA_OPERATE_OPENSEARCH_NUMBEROFSHARDS>
<CAMUNDA_OPERATE_OPENSEARCH_NUMBEROFREPLICAS>2</CAMUNDA_OPERATE_OPENSEARCH_NUMBEROFREPLICAS><!-- adjust based on number of nodes; test failures will tell you -->
<CAMUNDA_OPERATE_OPENSEARCH_AWS_ENABLED>true</CAMUNDA_OPERATE_OPENSEARCH_AWS_ENABLED>
<CAMUNDA_OPERATE_OPENSEARCH_AWS_REGION>${region}</CAMUNDA_OPERATE_OPENSEARCH_AWS_REGION>
<CAMUNDA_OPERATE_OPENSEARCH_AWS_SERVICENAME>es</CAMUNDA_OPERATE_OPENSEARCH_AWS_SERVICENAME>
```

Similarly, I adjusted the Zeebe config in [TestContainerUtil#startZeebe](https://github.com/camunda/camunda/blob/8.6.4/operate/qa/util/src/main/java/io/camunda/operate/qa/util/TestContainerUtil.java#L512-L529):

```java
.withEnv(
    "ZEEBE_BROKER_EXPORTERS_OPENSEARCH_ARGS_URL",
    "<aws url>")
.withEnv("AWS_ACCESS_KEY_ID", "<access key>")
.withEnv("AWS_SECRET_ACCESS_KEY", "<secret key>")
.withEnv("ZEEBE_BROKER_EXPORTERS_OPENSEARCH_ARGS_AWS_ENABLED", "true")
.withEnv("ZEEBE_BROKER_EXPORTERS_OPENSEARCH_ARGS_AWS_SERVICENAME", "es")
.withEnv("ZEEBE_BROKER_EXPORTERS_OPENSEARCH_ARGS_AWS_REGION", "<region>")
.withEnv("ZEEBE_BROKER_EXPORTERS_OPENSEARCH_ARGS_INDEX_NUMBEROFSHARDS", "1")
.withEnv("ZEEBE_BROKER_EXPORTERS_OPENSEARCH_ARGS_INDEX_NUMBEROFREPLICAS", "2") // adjust to number of nodes
.withEnv("ZEEBE_BROKER_EXPORTERS_OPENSEARCH_ARGS_BULK_DELAY", "1")
.withEnv("ZEEBE_BROKER_EXPORTERS_OPENSEARCH_ARGS_BULK_SIZE", "1")
.withEnv(
    "ZEEBE_BROKER_EXPORTERS_OPENSEARCH_CLASSNAME",
    "io.camunda.zeebe.exporter.opensearch.OpensearchExporter")
```

**Comment by ThorbenLindhauer** [comment]
Next I will raise PRs against main/8.6/8.5/8.4

**Comment by ThorbenLindhauer** [comment]
@houssain-barouni I have raised all the necessary PRs to change from the Apache Http client to the AWS CRT client.

Could you please follow up on:

* getting them reviewed and merged
* once merged, get QA to make a smoke test with (I suggest) 8.6 to verify that the applications generally work with AWS Opensearch and that the scroll log is gone

**Comment by houssain-barouni** [comment]
I also run Tasklist integration tests against AWS Opensearch
[tasklist-integration.txt](https://github.com/user-attachments/files/18226484/tasklist-integration.txt)

The same tests fail with the apache client, so we don't expect a regression introduce with the change of the client implementation

**Comment by houssain-barouni** [comment]
>once merged, get QA to make a smoke test with (I suggest) 8.6 to verify that the applications generally work with AWS Opensearch and that the scroll log is gone

I [discussed](https://camunda.slack.com/archives/C06BN4SRB6C/p1734687664191379) with @mschoe , smoke (maybe automated) tests will be run before Jan 7, 2025

**Comment by esraagamal6** [comment]
Manual Smoke test passed for Tasklist and Operate on a docker environment
cc: @houssain-barouni @ArezooAzar

**Comment by esraagamal6** [comment]
ℹ️ We faced below issues while testing the env with docker:
- Tasklist is not populating the userTasks
- Operate is not displaying the incidents

**Comment by ArezooAzar** [comment]
Here are the logs for Operate, Tasklist, and Zeebe, in case they are needed.
[zeebeLog.txt](https://github.com/user-attachments/files/18237843/zeebeLog.txt)
[tasklistLog.txt](https://github.com/user-attachments/files/18237844/tasklistLog.txt)
[OperateLog.txt](https://github.com/user-attachments/files/18237845/OperateLog.txt)

**Comment by houssain-barouni** [comment]
> ℹ️ We faced below issues while testing the env with docker:
> 
> * Tasklist is not populating the userTasks
> * Operate is not displaying the incidents

@esraagamal6 @ArezooAzar 
For Tasklist, it looks like it is a regression that we have in 8.6 but unrelated to the this issue.
For Operate, @ArezooAzar can you give me a scenario to reproduce the issue (not displaying incidents)

**Comment by esraagamal6** [comment]
@houssain-barouni After retesting tasklist with `registry.camunda.cloud/team-hto/tasklist:branch-stable-8-6` i can still see the scroll issue `io.camunda.tasklist.util.OpenSearchUtil.scroll(OpenSearchUtil.java:446) ~[tasklist-els-schema-8.6.7-SNAPSHOT.jar:8.6.7-SNAPSHOT]` when doing any of the below actions:
1. Scrolling in taskist tasks
2. Scrolling in list of task variables

[TasklistLogsScroll.docx](https://github.com/user-attachments/files/18318751/TasklistLogsScroll.docx)

https://github.com/user-attachments/assets/9e14f8c7-5d85-4347-8eea-397e6ed5c8a7


https://github.com/user-attachments/assets/c005619a-2dca-45a4-8d1e-e10a347e3fe1

**Comment by ArezooAzar** [comment]
@houssain-barouni :Today with @esraagamal6  we test operate and see the Scroll Error Log Issue:

- During migration testing, a scroll error log appeared. Unfortunately, we could not identify which specific scroll caused the issue.
- Efforts to reproduce the error were unsuccessful.
here  is the log 
[OperateScrollLogs.txt](https://github.com/user-attachments/files/18319956/OperateScrollLogs.txt)

**Comment by ThorbenLindhauer** [comment]
@ArezooAzar what was your test setup? Were only Operate and Zeebe running, or also Tasklist/Optimize?

**Comment by houssain-barouni** [comment]
@esraagamal6 @ArezooAzar the issue that you found is related to https://github.com/camunda/camunda/issues/26571, we will provide a fix for it

**Comment by houssain-barouni** [comment]
@esraagamal6 @ArezooAzar , the fix is merged on stable/8.6, can you docker pull the latest `registry.camunda.cloud/team-hto/tasklist:branch-stable-8-6` and test again ?

**Comment by ArezooAzar** [comment]
> @ArezooAzar what was your test setup? Were only Operate and Zeebe running, or also Tasklist/Optimize?

@ThorbenLindhauer : we used the docker file that @houssain-barouni  provided and it contain zeebe- Operate-tasklist. 
we run in order first Zeebe and Operate and then Tasklist

**Comment by ArezooAzar** [comment]
@ThorbenLindhauer and @houssain-barouni: @esraagamal6  and I tested the issue and found no scroll errors in Operate and TaskList. We also ran the smock test. Everything works as expected.
